debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java

```diff
@@ -570,7 +570,7 @@ private PGReplicationStream startPgReplicationStream(final Lsn lsn, Function<Cha
                 .getReplicationAPI()
                 .replicationStream()
                 .logical()
-                .withSlotName(slotName)
+                .withSlotName("\"" + slotName + "\"")
```

Comment:
I am wondering whether there's a compatibility concern here. Replication slots are persistent across connector restarts (so to prevent the DB from dropping WAL segments not consumed by a given slot). If a connector (existing version) has created a slot using the unquoted name, would we be able to resume that slot now using the quoted name?

This won't show up in our automated test suits, as we don't have that situation that a slot created by an earlier connector version is resumed. Hard to have an automated test for this, but a manual verification of this would suffice IMO.

Reply:
IMHO should be fine (this could be an issue in case of numeric slot names, but it doesn't work with current version). Anyway, definitely worth checking - will do (may me take some time).

Reply:
I suppose the simplest would be to take the [tutorial example](https://github.com/debezium/debezium-examples/tree/main/tutorial#using-postgres) and

* register a connector with the current version (1.7)
* build the Debezium Postgres connector with this change (`mvn clean verify -am -Dquick -Passembly`)
* stop the "connect" service
* update the connect image (or provide the connector via a mounted volume)
* restart and verify a) that the connector can continue with the previously created slot, though referring to it with quotes now, and b) that a new connector instance with a numeric slot name can be created (so to verify you pulled in the updated connector version)

Happy to help if you run into any issues. Thanks!

Reply:
Thanks for the validation steps!

AFAICT everything works:
* started compose with 1.7 and `"slot.name": "dbz_1042"`
* provided 1.8 snapshort with this change via mount dir
* restarted `connect` service
* CDC works after restart, relevant log part after restart:
```
connect_1    | 2021-11-17 21:21:57,764 INFO   Postgres|dbserver1|streaming  Obtained valid replication slot ReplicationSlot [active=false, latestFlushedLsn=LSN{0/2081100}, catalogXmin=601]   [io.debezium.connector.postgresql.connection.PostgresConnection]
connect_1    | 2021-11-17 21:21:57,767 INFO   Postgres|dbserver1|streaming  Connection gracefully closed   [io.debezium.jdbc.JdbcConnection]
postgres_1   | 2021-11-17 21:21:57.777 GMT [151] LOG:  starting logical decoding for slot "dbz_1042"
postgres_1   | 2021-11-17 21:21:57.777 GMT [151] DETAIL:  Streaming transactions committing after 0/2081100, reading WAL from 0/20810C8.
postgres_1   | 2021-11-17 21:21:57.777 GMT [151] STATEMENT:  START_REPLICATION SLOT dbz_1042 LOGICAL 0/2081100
postgres_1   | 2021-11-17 21:21:57.777 GMT [151] LOG:  logical decoding found consistent point at 0/20810C8
postgres_1   | 2021-11-17 21:21:57.777 GMT [151] DETAIL:  There are no running transactions.
postgres_1   | 2021-11-17 21:21:57.777 GMT [151] STATEMENT:  START_REPLICATION SLOT dbz_1042 LOGICAL 0/2081100
connect_1    | 2021-11-17 21:21:57,805 INFO   Postgres|dbserver1|streaming  Requested thread factory for connector PostgresConnector, id = dbserver1 named = keep-alive   [io.debezium.util.Threads]
connect_1    | 2021-11-17 21:21:57,806 INFO   Postgres|dbserver1|streaming  Creating thread debezium-postgresconnector-dbserver1-keep-alive   [io.debezium.util.Threads]
```

After compose stop/start, still with 1.8 snapshot and  numeric slot name `"slot.name": "1042"`, it starts and CDC works, log:
```
connect_1    | 2021-11-17 21:37:53,786 INFO   Postgres|dbserver1|streaming  Obtained valid replication slot ReplicationSlot [active=false, latestFlushedLsn=LSN{0/2078D38}, catalogXmin=601]   [io.debezium.connector.postgresql.connection.PostgresConnection]
connect_1    | 2021-11-17 21:37:53,787 INFO   Postgres|dbserver1|streaming  Connection gracefully closed   [io.debezium.jdbc.JdbcConnection]
kafka_1      | 2021-11-17 21:37:53,790 - INFO  [data-plane-kafka-request-handler-2:Logging@66] - [Partition dbserver1.inventory.geom-0 broker=1] No checkpointed highwatermark is found for partition dbserver1.inventory.geom-0
kafka_1      | 2021-11-17 21:37:53,790 - INFO  [data-plane-kafka-request-handler-2:Logging@66] - [Partition dbserver1.inventory.geom-0 broker=1] Log loaded for partition dbserver1.inventory.geom-0 with initial high watermark 0
postgres_1   | 2021-11-17 21:37:53.791 GMT [89] LOG:  starting logical decoding for slot "1042"
postgres_1   | 2021-11-17 21:37:53.791 GMT [89] DETAIL:  Streaming transactions committing after 0/2078D38, reading WAL from 0/2078D00.
postgres_1   | 2021-11-17 21:37:53.791 GMT [89] STATEMENT:  START_REPLICATION SLOT "1042" LOGICAL 0/2078D38
postgres_1   | 2021-11-17 21:37:53.791 GMT [89] LOG:  logical decoding found consistent point at 0/2078D00
postgres_1   | 2021-11-17 21:37:53.791 GMT [89] DETAIL:  There are no running transactions.
postgres_1   | 2021-11-17 21:37:53.791 GMT [89] STATEMENT:  START_REPLICATION SLOT "1042" LOGICAL 0/2078D38
connect_1    | 2021-11-17 21:37:53,804 INFO   Postgres|dbserver1|streaming  Requested thread factory for connector PostgresConnector, id = dbserver1 named = keep-alive   [io.debezium.util.Threads]
connect_1    | 2021-11-17 21:37:53,804 INFO   Postgres|dbserver1|streaming  Creating thread debezium-postgresconnector-dbserver1-keep-alive   [io.debezium.util.Threads]
```

For completeness (reproducing original issue), another compose stop/star with `"slot.name": "1042"` and 1.7, it fails with syntax error:
```
connect_1    | 2021-11-17 21:49:36,227 INFO   Postgres|dbserver1|postgres-connector-task  No previous offset found   [io.debezium.connector.postgresql.PostgresConnectorTask]
connect_1    | 2021-11-17 21:49:36,227 INFO   Postgres|dbserver1|postgres-connector-task  Taking initial snapshot for new datasource   [io.debezium.connector.postgresql.snapshot.InitialSnapshotter]
connect_1    | 2021-11-17 21:49:36,236 INFO   Postgres|dbserver1|postgres-connector-task  Creating replication slot with command CREATE_REPLICATION_SLOT 1042  LOGICAL decoderbufs   [io.debezium.connector.postgresql.connection.PostgresReplicationConnection]
postgres_1   | 2021-11-17 21:49:36.236 GMT [88] ERROR:  syntax error
connect_1    | 2021-11-17 21:49:36,238 ERROR  ||  WorkerSourceTask{id=inventory-connector-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted   [org.apache.kafka.connect.runtime.WorkerTask]
connect_1    | io.debezium.DebeziumException: Creation of replication slot failed
connect_1    |  at io.debezium.connector.postgresql.PostgresConnectorTask.start(PostgresConnectorTask.java:144)
connect_1    |  at io.debezium.connector.common.BaseSourceTask.start(BaseSourceTask.java:133)
connect_1    |  at org.apache.kafka.connect.runtime.WorkerSourceTask.initializeAndStart(WorkerSourceTask.java:231)
connect_1    |  at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:186)
connect_1    |  at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:237)
connect_1    |  at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
connect_1    |  at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
connect_1    |  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
connect_1    |  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
connect_1    |  at java.base/java.lang.Thread.run(Thread.java:829)
connect_1    | Caused by: org.postgresql.util.PSQLException: ERROR: syntax error
connect_1    |  at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2552)
connect_1    |  at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2284)
connect_1    |  at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:322)
connect_1    |  at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:481)
connect_1    |  at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:401)
connect_1    |  at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:322)
connect_1    |  at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:308)
connect_1    |  at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:284)
connect_1    |  at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:279)
connect_1    |  at io.debezium.connector.postgresql.connection.PostgresReplicationConnection.createReplicationSlot(PostgresReplicationConnection.java:363)
connect_1    |  at io.debezium.connector.postgresql.PostgresConnectorTask.start(PostgresConnectorTask.java:137)
connect_1    |  ... 9 more
```

Reply:
Ok, excellent. Thanks for the detailed verification 🙏 !